state/upgrades.go

```diff
@@ -3871,3 +3871,42 @@ func CleanupDeadAssignUnits(pool *StatePool) error {
 		return nil
 	}))
 }
+
+// RemoveOrphanedLinkLayerDevices removes link-layer devices and addresses
+// that have no corresponding machine in the model.
+// This situation could occur in the past for force-destroyed machines.
+func RemoveOrphanedLinkLayerDevices(pool *StatePool) error {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		machines, mCloser := st.db().GetCollection(machinesC)
+		defer mCloser()
+		iter := machines.Find(nil).Iter()
+
+		machineIDs := set.NewStrings()
+		var mDoc struct {
+			ID string `bson:"machineid"`
+		}
+		for iter.Next(&mDoc) {
+			machineIDs.Add(mDoc.ID)
+		}
+
+		if err := iter.Close(); err != nil {
+			return errors.Trace(err)
+		}
+
+		linkLayerDevices, lldCloser := st.db().GetCollection(linkLayerDevicesC)
+		defer lldCloser()
+		iter = linkLayerDevices.Find(nil).Iter()
+
+		var devDoc linkLayerDeviceDoc
+		for iter.Next(&devDoc) {
+			if machineIDs.Contains(devDoc.MachineID) {
+				continue
+			}
+			if err := newLinkLayerDevice(st, devDoc).Remove(); err != nil {
+				return errors.Trace(err)
```

Comment:
You need a `_ = iter.Close()` here to avoid leaking the iterator when returning with an error.